dsd/constraints.py

```diff
@@ -616,17 +644,48 @@ def generate_sequence(self, rng: np.random.Generator) -> str:
 
         :param rng:
             numpy random number generator to use. To use a default, pass :py:data:`np.default_rng`.
+        :param previous_sequence:
+            previously generated sequence to be replaced by a new sequence; None if no previous
+            sequence exists. Used in :py:meth:`DomainPool.find_steps_distance_sequences`
+            to choose a new sequence "close" to itself in Hamming distance. The number of
+            differences between previous_sequence and its neighbors is determined by randomly
+            picking a Hamming distance from :py:data:`DomainPool.hamming_probability` with
+            weighted probabilities of choosing each distance.
         :return:
             DNA sequence of given length satisfying :py:data:`DomainPool.numpy_constraints` and
             :py:data:`DomainPool.sequence_constraints`
         """
         log_debug_sequence_constraints_accepted = False
-        sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        while not self.satisfies_sequence_constraints(sequence):
-            logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+        if not self.replace_with_close_sequences or not previous_sequence:
+            # takes a completely random sequence from domain pool
             sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
-        if log_debug_sequence_constraints_accepted:
-            logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+            while not self.satisfies_sequence_constraints(sequence):
+                logger.debug(f'rejecting domain sequence {sequence}; failed some sequence constraint')
+                sequence = self._get_next_sequence_satisfying_numpy_constraints(rng)
+            if log_debug_sequence_constraints_accepted:
+                logger.debug(f'accepting domain sequence {sequence}; passed all sequence constraints')
+        else:
+            # takes neighbor to previous sequence; difference in bases randomly chosen
+            hamming_distances = []
+            hamming_probabilities = []
+            for key, val in self.hamming_probability.items(): # make lists for np.random.choice
+                hamming_distances.append(key)
+                hamming_probabilities.append(val)
+            # chooses random number of bases to be changed (steps)
+            steps = np.random.choice(hamming_distances, p=hamming_probabilities) # seed?
+            neighbors = self.find_steps_distance_sequences(steps, previous_sequence)
+            while len(neighbors) == 0:
```

Comment:
This loop makes me nervous since it could run forever. It also seems inefficient. What happens when `self._sequences` gets down to only a few sequences? If there's 5 of them, then if the domains are length 15, then at least 10 our of 15 possible values of `steps` will have no sequences that distance from `previous_sequence`. So we'll spend a lot of time searching the list.

A better solution would be for `find_steps_distance_sequences` to return a dict mapping each Hamming distance to a list of remaining sequences that are that Hamming distance from `previous_sequence`. You're already looping over all of the remaining sequences, so this won't take much more time. Then you only need to call `find_steps_distance_sequences` once. But it should be re-titled, since it won't take `steps` as a parameter anymore. Instead, you'll get that dict, then randomly select one of the keys in it as the value of `steps`.

Reply:
Hmm I understand the dict idea but how would the steps be chosen then? Because currently the number of steps uses weighted probabilities from the other dictionary, but if one of the Hamming distances isn't a key, then its probability of being chosen is zero and the other probability weights need to change to add to 1. I can make a function that adjusts the weights depending on how many keys there are (using decay by 1/2), but then the user wouldn't be able to specify probability weights as a dictionary, they would need to rewrite the probability weight function themselves. What do you suggest?

Reply:
Can we meet at 9am Friday to discuss?

Reply:
Sure, let's do that.